mltensor-explainer.md

```diff
@@ -0,0 +1,422 @@
+  const tensorizedGpuBufferAfterInference = await gpuDevice.importExternalBuffer(mlTensor2);
+
+  // Create a bind group for `tensorizedGpuBufferAfterInference`,
+  // create a command encoder, etc to feed `tensorizedGpuBufferAfterInference`
+  // into a GPU shader which may blur the frame or replace background sections
+  // and then render the result
+  // ...
+
+  gpuDevice.queue.submit([texturizeAndRenderCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN for the next frame.
+  tensorizedGpuBufferAfterInference.release();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines) for more details.
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The WebNN API requires the developer to declare how an `MLTensor` will be used (via `MLTensorDescriptor`), which the user agent may use as a hint in deciding where to allocate the memory backing an `MLTensor`. Where the memory is ultimately allocated is up to the user agent.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorDescriptor.importableToWebGPU` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorDescriptor.readable` and `MLTensorDescriptor.writable` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
+
+### Importing an `MLTensor` to WebGPU
+
+Any `MLTensor` created with the `MLTensorDescriptor.importableToWebGPU` flag may be imported into any `GPUDevice`. In the best case, this requires no data copies. If the underlying buffer backing the `MLTensor` is not accessible to the `GPUDevice`, this will require copying the contents of the `MLTensor` to a new buffer, then copying the contents of this buffer back to the `MLTensor` once WebGPU releases its handle to the buffer.
+
+While an `MLTensor` is rented to a `GPUDevice`, the `GPUDevice` has exclusive, read/write access to the imported buffer, which is created as a `GPUExternalBuffer` with `GPUBufferUsageFlags.STORAGE`. All WebNN work depending - directly or indirectly - on the imported `MLTensor` is blocked until the `GPUDevice` returns the tensor.
```

Comment:
Sorry if this has been answered earlier: is the size and layout of the buffer defined somewhere such that the WGSL code manipulating the buffer knows where to find each element?

Reply:
Please correct me if I'm misunderstanding the question, but yes, I believe so. The size of the buffer is determined by its [data type and shape](https://www.w3.org/TR/webnn/#dictdef-mloperanddescriptor). The layout is an implementation detail from the perspective of WebNN, but in practice the buffer will always be a type that GPU code (in the Chromium implementation, at least) understands. For example on Mac we'll import the buffer as an `IOSurface` (which can be imported [as a shared image](https://crsrc.org/c/gpu/command_buffer/service/shared_image/iosurface_image_backing.h))

Reply:
Ok I think we need more precision here. How would a user write a WGSL shader that manipulates a tensor when trying to for example implement custom operators? WebGPU sees storage buffers as a byte-addressed linear allocation, so the application needs to be aware of the layout of the tensor to address the allocation to find a specific element.

Reply:
Added a snippet showing the WGSL declaration of the imported buffer. See the diff in the latest patch: https://github.com/webmachinelearning/webnn/pull/754/commits/4af9354e433a871d48a3845177a767190b6c9696